base/compiler/typeinfer.jl

```diff
@@ -709,6 +709,7 @@ function merge_call_chain!(parent::InferenceState, ancestor::InferenceState, chi
     # and merge all of the callers into ancestor.callers_in_cycle
     # and ensure that walking the parent list will get the same result (DAG) from everywhere
     while true
+        parent = parent::InferenceState
```

Comment:
nit: might make more sense to put this on the `child.parent` line?

Reply:
I tried it first, but it turns out that this line can be `child.parent::Nothing` (will escape the loop right after that, though), so type assertion here

Reply:
I tried it first, but it turns out that this line can be `child.parent::Nothing` (will escape the loop right after that), so type assertion here

Reply:
alternatively we could explicit check against that case:
```julia
function merge_call_chain!(parent::InferenceState, ancestor::InferenceState, child::InferenceState)
    # add backedge of parent <- child
    # then add all backedges of parent <- parent.parent
    # and merge all of the callers into ancestor.callers_in_cycle
    # and ensure that walking the parent list will get the same result (DAG) from everywhere
    while true
        add_cycle_backedge!(child, parent, parent.currpc)
        union_caller_cycle!(ancestor, child)
        child = parent
        parent = child.parent
        parent === nothing && break
        child === ancestor && break
    end
end
```

Reply:
ah, or in that case:
```
        child === ancestor && break
        parent = child.parent::InferenceState
```

Reply:
That should be better. Addressed.